### PR TITLE
fast tracked append

### PR DESCRIPTION
**Pull request by porkybrain**

### CODE REVIEW

**Comment by porkybrain** [comment]
looks like we could implement `remove_if` quite fast if we had a fast append as mentioned [here](https://github.com/edouarda/brigand/issues/127#issuecomment-190306110) so I would like to incite a discussion about appending here.

on this benchmark:
```C++
using l = list<list<i<0>, i<1000>>, ... 1000 elements ..., list<i<999>, i<1999>>>;
using j = brigand::wrap<l,brigand::append>;
```
I compared my [super_join](https://github.com/edouarda/brigand/issues/127#issuecomment-190306110) with the current append and a 16x fast tracked append (I subtracted the compile time of the same project with the append line commented out so this is just the append it self):
super_join: 0.25s
fast tracked append: 0.25s
append :3.33s


In conclusion deducing the list type of every list in the fast track (difference between fast tracked join and super_join) is a lot of typing but costs nothing.

**Comment by porkybrain** [comment]
Benchmarks for brigand::remove optimization 450 element list and removing 40 different elements using clang

old brigand implementation: 11.55s

fast track raking 3 at a time old brigand: 6.93s  
(fast tracking with more than 3 sucks because typing is quadratic)

[origionally preposed algorithm](https://github.com/edouarda/brigand/issues/127#issuecomment-190306110): 1.30s

the committed version: 0.8s

**Comment by porkybrain** [comment]
sorry for being stupid ;) but good thing I put in those custom list uni tests.

File: brigand/algorithms/remove.hpp [hunk 1]
```diff
@@ -6,51 +6,53 @@
...
+#include <brigand/sequences/append.hpp>
+#include <brigand/sequences/clear.hpp>
 #include <type_traits>
 
 namespace brigand
 {
-  namespace detail
-  {
-    template< class L1, typename Pred, class L2 = clear<L1>>
+namespace detail
```
Comment (jfalcou) [inline brigand/algorithms/remove.hpp]: could be cool if those spacing change didn't get in as they make a lot of noise and derail PR review.

Comment (porkybrain) [inline brigand/algorithms/remove.hpp]: Yeah I probably should have only run clang format on the part that I changed rather than the whole file. I'll fix it tomorrow (on my phone now)
